fpdf/fonts.py

```diff
@@ -87,67 +107,219 @@ def __init__(self, fpdf, font_file_path, fontkey, style):
+        self.ttfont.close()
+        self.hbfont = None
+
+    def get_text_width(self, text, font_size_pt, text_shaping_parms):
+        if text_shaping_parms:
+            return self.shaped_text_width(text, font_size_pt, text_shaping_parms)
+        return (len(text), sum(self.cw[ord(c)] for c in text) * font_size_pt * 0.001)
+
+    def shaped_text_width(self, text, font_size_pt, text_shaping_parms):
+        """
+        When texts are shaped, the length of a string is not always the sum of all individual character widths
+        This method will invoke harfbuzz to perform the text shaping and return the sum of "x_advance"
+        and "x_offset" for each glyph. This method works for "left to right" or "right to left" texts.
+        """
+        _, glyph_positions = self.perform_harfbuzz_shaping(
+            text, font_size_pt, text_shaping_parms
+        )
+        text_width = 0
+        for pos in glyph_positions:
+            text_width += (
+                round(self.scale * (pos.x_advance + pos.x_offset) + 0.001)
+                * font_size_pt
+                * 0.001
+            )
+        return (len(glyph_positions), text_width)
+
+    # Disabling this check - looks like cython confuses pylint:
+    # pylint: disable=no-member
+    def perform_harfbuzz_shaping(self, text, font_size_pt, text_shaping_parms):
+        """
+        This method invokes Harfbuzz to performing text shaping of the input string
+        """
+        if not hasattr(self, "hbfont"):
+            self.hbfont = hb.Font(hb.Face(hb.Blob.from_file_path(self.ttffile)))
+        self.hbfont.ptem = font_size_pt
+        buf = hb.Buffer()
+        buf.cluster_level = 1
+        buf.add_str("".join(text))
+        features = text_shaping_parms["features"]
+        if (
+            text_shaping_parms["direction"]
+            or text_shaping_parms["script"]
+            or text_shaping_parms["language"]
+        ):
+            buf.direction = text_shaping_parms["direction"]
+            buf.script = text_shaping_parms["script"]
+            buf.language = text_shaping_parms["language"]
+        else:
+            buf.guess_segment_properties()
+        hb.shape(self.hbfont, buf, features)
+        return buf.glyph_infos, buf.glyph_positions
+
+    def encode_text(self, text):
+        txt_mapped = ""
+        for char in text:
+            uni = ord(char)
+            # Instead of adding the actual character to the stream its code is
+            # mapped to a position in the font's subset
+            txt_mapped += chr(self.subset.pick(uni))
+        return f'({escape_parens(txt_mapped.encode("utf-16-be").decode("latin-1"))}) Tj'
+
+    def shape_text(self, text, font_size_pt, text_shaping_parms):
+        """
+        This method will invoke harfbuzz for text shaping, include the mapping code
+        of the glyphs on the subset and map input characters to the cluster codes
+        """
+        if len(text) == 0:
+            return zip([], [])
```

Comment:
```suggestion
            return []
```
This would make the code slightly clearer, and avoid returning a `zip object` that is not exactly a list